src/blockaxis.jl

```diff
@@ -16,67 +16,188 @@ function findblockindex(b::AbstractVector, k::Integer)
+    prevblocklast = blockidx == firstindex(bl) ? first(b)-oneunit(eltype(b)) : bl[blockidx-1]
     local_index = k - prevblocklast
     return BlockIndex(blockidx, local_index)
 end
 
+abstract type AbstractBlockedUnitRange{T,CS} <: AbstractUnitRange{T} end
+
 function _BlockedUnitRange end
 
 
 """
     BlockedUnitRange
 
-is an `AbstractUnitRange{Int}` that has been divided
-into blocks, and is used to represent axes of block arrays.
+is an `AbstractUnitRange{Int}` that has been divided into blocks.
 Construction is typically via `blockedrange` which converts
 a vector of block lengths to a `BlockedUnitRange`.
 
 # Examples
 ```jldoctest
-julia> blockedrange([2,2,3])
-3-blocked 7-element BlockedUnitRange{Vector{Int64}}:
- 1
+julia> blockedrange(2, [2,2,3]) # first value and block lengths
+3-blocked 7-element BlockedUnitRange{Int64, Vector{Int64}}:
  2
- ─
  3
- 4
  ─
+ 4
  5
+ ─
  6
  7
+ 8
 ```
+
+See also [`BlockedOneTo`](@ref).
 """
-struct BlockedUnitRange{CS} <: AbstractUnitRange{Int}
-    first::Int
+struct BlockedUnitRange{T<:Integer,CS} <: AbstractBlockedUnitRange{T,CS}
+    first::T
     lasts::CS
-    global _BlockedUnitRange(f, cs::CS) where CS = new{CS}(f, cs)
+    # assume that lasts is sorted, no checks carried out here
+    global function _BlockedUnitRange(f::T, cs::CS) where {T,CS<:AbstractVector{T}}
+        Base.require_one_based_indexing(cs)
+        return new{T,CS}(f, cs)
+    end
+    global function _BlockedUnitRange(f::T, cs::CS) where {T,CS<:Tuple{T,Vararg{T}}}
+        return new{T,CS}(f, cs)
+    end
+    global function _BlockedUnitRange(f::T, cs::Tuple{}) where {T}
+        return new{T,Tuple{}}(f, cs)
+    end
 end
 
-const DefaultBlockAxis = BlockedUnitRange{Vector{Int}}
-
-@inline _BlockedUnitRange(cs) = _BlockedUnitRange(1,cs)
+@inline function _BlockedUnitRange(f::T, cs::AbstractVector{S}) where {T,S}
+  U = promote_type(T, S)
+  return _BlockedUnitRange(convert(U, f), convert.(U, cs))
+end
+@inline function _BlockedUnitRange(f::T, cs::Tuple{S,Vararg{S}}) where {T,S}
+  U = promote_type(T, S)
+  return _BlockedUnitRange(convert(U, f), convert.(U, cs))
+end
+@inline function _BlockedUnitRange(f, cs::Tuple)
+  return _BlockedUnitRange(f, promote(cs...))
+end
+@inline _BlockedUnitRange(cs::AbstractVector) = _BlockedUnitRange(oneunit(eltype(cs)), cs)
+@inline _BlockedUnitRange(cs::NTuple) = _BlockedUnitRange(oneunit(eltype(cs)), cs)
+_BlockedUnitRange(cs::Tuple) = _BlockedUnitRange(promote(cs...))
 
+first(b::BlockedUnitRange) = b.first
+@inline blocklasts(a::BlockedUnitRange) = a.lasts
 
 BlockedUnitRange(::BlockedUnitRange) = throw(ArgumentError("Forbidden due to ambiguity"))
-_blocklengths2blocklasts(blocks) = cumsum(blocks) # extra level to allow changing default cumsum behaviour
-@inline blockedrange(blocks::Union{Tuple,AbstractVector}) = _BlockedUnitRange(_blocklengths2blocklasts(blocks))
+# Use `accumulate` instead of `cumsum` because it preserves the element type of the block lengths
```

Comment:
@jishnub  I don't understand this change. Can you give an example where they are different?

Reply:
This also will break a lot of smart block ranges eg:
```
julia> cumsum(Fill(2,10))
2:2:20

julia> accumulate(+,Fill(2,10))
10-element Vector{Int64}:
  2
  4
  6
  8
 10
 12
 14
 16
 18
 20
```

Reply:
Ah ok I tried reverting the change and I see the issue.

I don't like the special cases below, instead we should specialise `accumulate(+, ::AbstractFill)`.